### PR TITLE
Improve GitHub Actions workflow

### LINKED ISSUE #24: Improve GitHub Actions workflow
### Description

Currently, the GitHub actions fail if a PR modifies any file apart from the config files. It shouldn't fail if a PR contains other necessary changes submitted by a team member.

### Acceptance Criteria

- The modified file whitelist validation must be bypassed if the PR author belongs to the development team
- All the validation checks are performed regardless of their success status
- All the errors must be collected and presented at the end of all the validation checks

### PR DESCRIPTION
**Pull request by vardan10**
### What was the problem?
This PR resolves #24

### How was it solved?

- [x] The modified file whitelist validation must be bypassed if the PR author belongs to the development team
- [x] All the validation checks are performed regardless of their success status
- [x] All the errors must be collected and presented at the end of all the validation checks
- [x] Update test cases

### How was it tested?

- [x] Locally using [act](https://github.com/nektos/act)

### CODE REVIEW

File: validation/src/index.js [hunk 1]
```diff
@@ -50,19 +52,23 @@ const validate = async () => {
...
 	// Check for config files in all changed apps in networks directories
-	await validateAllConfigFiles(changedAppDirs);
+	await validateAllConfigFiles(changedAppDirs, validationErrors);
 
 	// Validate serviceURLs
-	await validateURLs(changedAppFiles);
+	await validateURLs(changedAppFiles, validationErrors);
+
+	if (validationErrors.length > 0) {
+		throw new Error(validationErrors);
```
Comment (sameersubudhi) [inline validation/src/index.js:70]:
```suggestion
		throw new Error(validationErrors.join('\n'));
```

File: validation/src/validateURLs.js [hunk 2]
```diff
@@ -92,35 +114,43 @@ const validateAppNodeUrls = async (appNodeInfos) => {
 		try {
 			await checkWebsocketConnection(appNodeUrl);
 		} catch (e) {
-			throw new Error(`Error establishing connection with node: ${appNodeUrl}`);
+			validationErrors.push(new Error(`Error establishing connection with node: ${appNodeUrl}`));
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:117]:
```suggestion
			validationErrors.push(new Error(`Error establishing connection with node: ${appNodeUrl}.`));
```

File: .github/workflows/pr-validation.yml [hunk 3]
```diff
@@ -6,6 +6,12 @@ jobs:
     steps:
       - name: Checkout code
         uses: actions/checkout@v3
+      - uses: tspascoal/get-user-teams-membership@v2
+        id: checkUserMember
+        with:
+          username: ${{ github.actor }}
+          team: 'LiskHQ'
```
Comment (sameersubudhi) [inline .github/workflows/pr-validation.yml:13]:
```suggestion
          organization: 'LiskHQ'
          team: 'platform'
```

**Review by sameersubudhi — COMMENTED** [review]
Shouldn't it be with both `organization` and `team` params?

File: .github/workflows/pr-validation.yml [hunk 4]
```diff
@@ -6,6 +6,13 @@ jobs:
     steps:
       - name: Checkout code
         uses: actions/checkout@v3
+      - uses: tspascoal/get-user-teams-membership@v2
+        id: checkUserMember
+        with:
+          username: ${{ github.actor }}
+          organization: 'LiskHQ'
+          team: 'Platform'
```
Comment (sameersubudhi) [inline .github/workflows/pr-validation.yml:14]:
```suggestion
          team: 'platform'
```

File: validation/src/index.js [hunk 5]
```diff
@@ -21,8 +21,13 @@ const { validateConfigFilePaths } = require('./validateConfigFilePaths');
 const config = require('../config');
 
 const validate = async () => {
+	const validationErrors = [];
+
+	// Check if organization author modified code
+	const doesAutherBelongToOrg = process.argv[2];
```
Comment (sameersubudhi) [inline validation/src/index.js:27]:
```suggestion
	// Check if the PR author from the @LiskHQ/platform team
	const isAuthorFromDevTeam = process.argv[2];
```

File: validation/src/index.js [hunk 6]
```diff
@@ -50,19 +55,25 @@ const validate = async () => {
...
 
 	// Validate serviceURLs
-	await validateURLs(changedAppFiles);
+	await validateURLs(changedAppFiles, validationErrors);
+
+	if (validationErrors.length > 0 && !doesAutherBelongToOrg) {
+		throw new Error(validationErrors.join('\n'));
+	} else if (validationErrors.length > 0) {
+		console.warn(validationErrors.join('\n'));
+	}
```
Comment (sameersubudhi) [inline validation/src/index.js:76]: We should throw errors whenever there's a validation error. But with `isAuthorFromDevTeam: true`, we can skip the `validateAllWhitelistedFiles` invocation on L64.

Also, please verify if `isAuthorFromDevTeam` is a boolean value of a string. Most likely, it is a string.

```suggestion
	if (validationErrors.length > 0) {
		throw new Error(validationErrors.join('\n'));
	}
```

File: validation/tests/validateConfigFilePaths.test.js [hunk 7]
```diff
@@ -28,21 +28,29 @@ describe('Schema Validation Tests', () => {
 		await setup.cleanTestEnviroment();
 	});
 
-	it('Should not throw if config files are present in network dir', async () => {
+	it('Should not have validation errors if config files are present in network dir', async () => {
```
Comment (sameersubudhi) [inline validation/tests/validateConfigFilePaths.test.js:31]:
```suggestion
	it('should not have validation errors if config files are present in network dir', async () => {
```

File: validation/tests/validateConfigFilePaths.test.js [hunk 8]
```diff
@@ -28,21 +28,29 @@ describe('Schema Validation Tests', () => {
...
+		await validateConfigFilePaths(setup.getJSONFilesFromNetwork(), validationErrors);
+		expect(validationErrors.length).toBe(0);
+
 		await setup.removeFileFromNetwork(config.filename.APP_JSON);
 		await setup.removeFileFromNetwork(config.filename.NATIVE_TOKENS);
 		/* eslint-enable max-len */
 	});
 
-	it('Should throw if config files are present in non network dir', async () => {
+	it('Should have validation errors if config files are present in non network dir', async () => {
```
Comment (sameersubudhi) [inline validation/tests/validateConfigFilePaths.test.js:45]:
```suggestion
	it('should have validation errors if config files are present in non network dir', async () => {
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: validation/src/validateURLs.js [hunk 9]
```diff
@@ -151,10 +181,12 @@ const validateURLs = async (files) => {
 		/* eslint-disable no-await-in-loop */
 		const data = await readJsonFile(nativetokenFile);
 
-		for (let j = 0; j < data.tokens.length; j++) {
-			const token = data.tokens[j];
-			// Validate logo URLs
-			await validateLogoUrls(token.logo);
+		if (data.tokens) {
+			for (let j = 0; j < data.tokens.length; j++) {
+				const token = data.tokens[j];
```
[the remark this answers is not in the document]
Comment (vardan10) [reply validation/src/validateURLs.js:186]: That would be my preferred way but its causing linting issues:
`iterators/generators require regenerator-runtime, which is too heavyweight for this guide to allow them. Separately, loops should be avoided in favor of array iterations`

[the remark this answers is not in the document]
Comment (priojeetpriyom) [reply validation/src/validateURLs.js:186]: We can suppress this error for the specific line like below:

File: validation/src/validateConfigFiles.js [hunk 10]
```diff
@@ -15,24 +15,26 @@
 const fs = require('fs').promises;
 const config = require('../config');
 
-const validateAllConfigFilesInDir = async (directory) => {
+const validateAllConfigFilesInDir = async (directory, validationErrors) => {
```
Comment (sameersubudhi) [inline validation/src/validateConfigFiles.js:18]: Same as above.

File: validation/src/validateURLs.js [hunk 11]
```diff
@@ -41,23 +41,45 @@ const wsRequest = async (wsEndpoint) => {
 	}
 };
 
-const validateExplorerUrls = async (explorers) => {
+const validateExplorerUrls = async (explorers, validationErrors) => {
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:44]: Same as above.

File: validation/src/validateURLs.js [hunk 12]
```diff
@@ -82,7 +104,7 @@ const checkWebsocketConnection = async (url) => {
 	});
 };
 
-const validateAppNodeUrls = async (appNodeInfos) => {
+const validateAppNodeUrls = async (appNodeInfos, validationErrors) => {
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:107]: Same as above.

File: validation/src/validateURLs.js [hunk 13]
```diff
@@ -92,35 +114,45 @@ const validateAppNodeUrls = async (appNodeInfos) => {
...
 		} catch (e) {
-			throw new Error(`Error establishing connection with node: ${appNodeUrl}`);
+			validationErrors.push(new Error(`Error establishing connection with node: ${appNodeUrl}.`));
 		}
 		/* eslint-enable no-await-in-loop */
 	}
 };
 
-const validateServiceURLs = async (serviceURLs, chainID) => {
+const validateServiceURLs = async (serviceURLs, chainID, validationErrors) => {
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:123]: Same as above.

**Review by sameersubudhi — CHANGES_REQUESTED** [review]
Looks good overall but lets stick to the following.

**Review by sameersubudhi — APPROVED** [review]

**Review by priojeetpriyom — APPROVED** [review]

**Pull request merged**